### PR TITLE
[WIP] RBFE optimization

### PR DESCRIPTION
**Pull request by proteneer**
Edit - okay this PR turned into an unfortunate mega PR:
- [x] Drastically improve performance of code close to that of the non-alchemical system
- [x] Pushes lambda argument into the bonded potentials
- [x] Updates topology handlers to generate a single fused set of potentials
- [x] Gets rid of the manual VJP system entirely  
- [x] Separates preparation of the system from execution of the system
- [x] Implements a differentiable estimator.py::deltaG function that allows jax to do the chain rule
- [x] Fixes a super annoying pickle bug with the InterpolatedPotential class (__getattr__).

This PR attempts to address some of the issues seen #326. So far the diagnosis seems to be:

We're reaching the limit of atomicAdd collisions, as ```k_reduce_add_force_buffer``` in lambda_potential.cu is taking up close to 9% of the total runtime. Part of this is because for every small molecule bonded term, its split into 3 parts (lhs, core, rhs) to implement the linear rescaling. It also dramatically increases the number of kernels required to deal with the extra buffering. 

So we're going to push the ```(1-lamb)*U0 and lamb*U1``` to the inner loop to see if it helps. This is backwards compatible, so we can modify the combining code for the SingleTopology class in a separate pass if needed.

i.e. This allows for U_bonded(x, lamb, params) to be a proper function of lambda. Where as before we relied on the LambdaPotential to do the transformation.

Some technical parts:
1) This will require a upgrade to C++17 (which is fortunately compatible with CUDA10+)
2) Pybind had to be upgrade to support std::optional in order to support default args in not super painful way
3) I will be slowly converting FLOAT_TO_FIXED code

Potentials
- [x] HarmonicBond
- [x] HarmonicAngle
- [x] PeriodicTorsion

### CODE REVIEW

File: tests/test_free_energy_v2.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                prod_steps
+            )
+
+            dG = estimator.deltaG(model, sys_params)
+            dGs.append(dG)
+
+        return dGs[0] - dGs[1]
+
+    # automatic chaining of vjps
+    vg_fn = jax.value_and_grad(binding_model)
```
Comment (proteneer) [inline tests/test_free_energy_v2.py:165]: @maxentile I just wanted to briefly point you to this glorious line of code

Comment (maxentile) [reply tests/test_free_energy_v2.py:165]: Nice! This could greatly simplify later steps, and make it easier to plug in different loss functions or likelihood functions down the road.

File: timemachine/potentials/bonded.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def restraint(conf, lamb, params, lamb_flags, box, bond_idxs):
     return energy
 
 # lamb is *not used* it is used in the alchemical stuff after
-def harmonic_bond(conf, params, box, lamb, bond_idxs):
+def harmonic_bond(conf, params, box, lamb, bond_idxs, lamb_mult=None, lamb_offset=None):
     """
```
Comment (maxentile) [inline timemachine/potentials/bonded.py:72]: Update Parameters in docstring to include

```
    lamb_mult: None, or broadcastable to bond_idxs.shape[0]
         prefactor = (lamb_offset + lamb_mult * lamb)

    lamb_offset: None, or broadcastable to bond_idxs.shape[0]
         prefactor = (lamb_offset + lamb_mult * lamb)
```

or similar

File: timemachine/potentials/bonded.py [hunk 3]
```diff
@@ -97,18 +97,22 @@ def harmonic_bond(conf, params, box, lamb, bond_idxs):
...
     r0s = params[:, 1]
-
-    energy = np.sum(kbs/2 * np.power(dij - r0s, 2.0))
+    prefactor = (lamb_offset + lamb_mult * lamb)
+    energy = np.sum(prefactor * kbs/2 * np.power(dij - r0s, 2.0))
     return energy
 
 
-def harmonic_angle(conf, params, box, lamb, angle_idxs, cos_angles=True):
+def harmonic_angle(conf, params, box, lamb, angle_idxs, lamb_mult=None, lamb_offset=None, cos_angles=True):
```
Comment (maxentile) [inline timemachine/potentials/bonded.py:115]: Update Parameters in docstring to include

```
    lamb_mult: None, or broadcastable to angle_idxs.shape[0]
         prefactor = (lamb_offset + lamb_mult * lamb)

    lamb_offset: None, or broadcastable to angle_idxs.shape[0]
         prefactor = (lamb_offset + lamb_mult * lamb)
```

or similar

File: timemachine/potentials/bonded.py [hunk 4]
```diff
@@ -216,7 +225,7 @@ def signed_torsion_angle(ci, cj, ck, cl):
     return np.arctan2(y, x)
 
 
-def periodic_torsion(conf, params, box, lamb, torsion_idxs):
+def periodic_torsion(conf, params, box, lamb, torsion_idxs, lamb_mult=None, lamb_offset=None):
```
Comment (maxentile) [inline timemachine/potentials/bonded.py:228]: Update Parameters in docstring to include

```
    lamb_mult: None, or broadcastable to torsion_idxs.shape[0]
         prefactor = (lamb_offset + lamb_mult * lamb)

    lamb_offset: None, or broadcastable to torsion_idxs.shape[0]
         prefactor = (lamb_offset + lamb_mult * lamb)
```

or similar

File: examples/rbfe_single.py [hunk 5]
```diff
@@ -236,14 +129,87 @@ def run_epoch(client, ff, mol_a, mol_b, core):
...
+
+    # client = None
+
+    binding_model = model.RBFEModel(
+        client,
+        ff,
+        complex_system,
+        complex_coords,
+        complex_box,
+        complex_schedule,
```
Comment (maxentile) [inline examples/rbfe_single.py:174]: Probably later want to group these

File: examples/rbfe_single.py [hunk 6]
```diff
@@ -236,14 +129,87 @@ def run_epoch(client, ff, mol_a, mol_b, core):
...
+
+    gradient_clip_thresholds = {
+        nonbonded.AM1CCCHandler: 0.05,
+        nonbonded.LennardJonesHandler: np.array([0.001,0])
+    }
+
+    for epoch in range(100):
 
         epoch_params = serialize_handlers(ff_handlers)
+        loss, loss_grad = vg_fn(ordered_params, mol_a, mol_b, core, label_ddG)
```
Comment (maxentile) [inline examples/rbfe_single.py:196]: NICE

File: fe/free_energy.py [hunk 7]
```diff
@@ -130,90 +74,42 @@ def __init__(self, mol, ff):
...
 
+        host_coords: np.array
+            Nx3 array of atomic coordinates
 
         Returns
         -------
-        float, float
-            Returns a pair of average du_dl values for bonded and nonbonded terms.
+        4 tuple
+            bound_potentials, vjp_fns_and_handles, combined_masses, combined_coords
```
Comment (maxentile) [inline fe/free_energy.py:96]: update to match return types
```suggestion
            final_potentials, final_params, combined_masses, combined_coords
```

File: fe/free_energy.py [hunk 8]
```diff
@@ -241,101 +136,53 @@ def _get_integrator(self, combined_masses):
...
-            Number of steps to run production. Statistics are gathered.
+        ff_params: tuple of np.array
+            Exactly equal to bond_params, angle_params, proper_params, improper_params, charge_params, lj_params
 
         Returns
         -------
-        float, float
-            Returns a pair of average du_dl values for bonded and nonbonded terms.
+        4 tuple
+            unbound_potentials, system_parameters, combined_masses, combined_coords
```
Comment (maxentile) [inline fe/free_energy.py:151]: update to match return types
```suggestion
            final_potentials, final_params, combined_masses, combined_coords
```

File: fe/topology.py [hunk 9]
```diff
@@ -15,76 +15,131 @@
...
+        return self._parameterize_bonded_term(guest_params, guest_potential, self.host_periodic_torsion)
 
     def parameterize_nonbonded(self, ff_q_params, ff_lj_params):
-        # this needs to take care of the case when there's parameter interpolation.
         num_guest_atoms = self.guest_topology.get_num_atoms()
         guest_qlj, guest_p = self.guest_topology.parameterize_nonbonded(ff_q_params, ff_lj_params)
-        
+
+        if isinstance(guest_p, potentials.InterpolatedPotential):
+            assert guest_qlj.shape[0] == num_guest_atoms*2
```
Comment (maxentile) [inline fe/topology.py:127]: Possibly suggests a slight refactoring here, to reduce amount of second-guessing whether `guest_p` is parameter-interpolated or not

**Review by maxentile — APPROVED** [review]
Nice! Left some minor comments, but overall I'm very happy to get this into main branch

File: timemachine/potentials/bonded.py (hunk 2 above)
Comment (proteneer) [reply timemachine/potentials/bonded.py:72]: will fix - i like the "broadcastable to bond_idxs.shape[0]" part

File: examples/rbfe_single.py (hunk 5 above)
Comment (proteneer) [reply examples/rbfe_single.py:174]: agreed!

File: fe/topology.py (hunk 9 above)
Comment (proteneer) [reply fe/topology.py:129]: cleaned this logic part up a bit

Comment (maxentile) [reply fe/topology.py:129]: Nice! Marking resolved

**Comment by proteneer** [comment]
Merging!
